src/persist-client/src/internal/state.rs

```diff
@@ -264,6 +265,30 @@ impl<T> BatchPart<T> {
         }
     }
 
+    pub fn structured_key_lower(&self) -> Option<ArrayBound> {
+        let part = match self {
+            BatchPart::Hollow(part) => part,
+            BatchPart::Inline { .. } => return None,
+        };
+
+        let try_decode = |lower: &LazyProto<ProtoArrayData>| {
+            let proto = lower.decode()?;
+            let data = ArrayData::from_proto(proto)?;
+            ensure!(data.len() == 1);
+            Ok(ArrayBound::new(make_array(data), 0))
+        };
+
+        let decoded: anyhow::Result<ArrayBound> = try_decode(part.structured_key_lower.as_ref()?);
+
+        match decoded {
+            Ok(bound) => Some(bound),
+            Err(e) => {
+                error!("failed to decode bound: {:#?}", e);
```

Comment:
After #29261 meant that we silently had a lot less test coverage than we thought, I get nervous about bare `error!` logs. IMO persist should basically always (always always?) be using one of the soft_assert_or_log variants instead of `error!`

Reply:
Yeah, that's better - thanks!